### ISSUE TITLE
Dynamic routes ignore segment revalidation settings

### ISSUE THREAD

**Issue by benderillo** [vercel/next.js#46978]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.6.0: Sat Jun 18 17:05:47 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.14.0
      npm: 8.5.0
      Yarn: 1.22.19
      pnpm: 6.11.0
    Relevant packages:
      next: 13.2.4-canary.9
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-ssxpmi?file=app/pages/[slug]/page.tsx

### To Reproduce

1. Build and start the project that reproduces the issue: https://stackblitz.com/edit/vercel-next-js-ssxpmi?file=app/pages/[slug]/page.tsx
2. Navigate to any of the pages (page1, page2 or page3) via the links on the home page.
3. Refresh the page
4. Observe the log in the terminal saying the page is re-rendered on each page fetch (SSR).

### Describe the Bug

**TL;DR**
Dynamic route segment rendering ignores the segment configuration settings when decides whether the page should be rendered dynamically or statically.

**Details**

Presence of `generateStaticParams` dictates whether the segment is dynamically or statically rendered, which is neither documented nor intuitive.

The segment configuration in the provided repo exports `export const revalidate = false` but the page is still re-rendered on each request.
And the expectation is that the page is generated on a first request and kept forever.

If I define `generateStaticParams()` function that returns empty array, then I get the expected behaviour: nothing is generated at build time and all my routes are generated and cached forever on the first visit.
 

### Expected Behavior

Segment configuration settings should be decoupled.

1. Presence of `generateStaticParams` should only control if there is any build-time pre-generation of dynamic segments should happen. Nothing more than that.

If the function is defined, NextJS should call it to get the list of segments for pre-generation at build time. No more extra logic assigned to it. If the list is empty -> nothing is pre-generated. If the function is not defined -> nothing is pre-generated.

2. What happens when there is visit of a dynamic segment that does not exist should be dictated by the `[dynamicParams](https://beta.nextjs.org/docs/api-reference/segment-config#dynamicparams)`. If it says "true" then the page is generated upon request.

3. What happens to the generated page on subsequent requests is controlled by the cache and revalidation settings or already defined heuristics around use of dynamic functions. (presence or absence of the `generateStaticParams` should not impact this).

### Which browser are you using? (if relevant)

chrome

### How are you deploying your application? (if relevant)

vercel
